File: web-dashboard/src/services/api-client.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import { FirebaseAuthService } from './firebase/auth.service';

export class ApiClient {
  private static instance: ApiClient;
  private axiosInstance: AxiosInstance;
  private authService: FirebaseAuthService;

  private constructor() {
    this.authService = FirebaseAuthService.getInstance();
    this.axiosInstance = axios.create({
      baseURL: process.env.REACT_APP_API_BASE_URL || 'http://localhost:8000',
      timeout: 10000,
    });

    this.setupInterceptors();
  }

  static getInstance(): ApiClient {
    if (!ApiClient.instance) {
      ApiClient.instance = new ApiClient();
    }
    return ApiClient.instance;
  }

  private setupInterceptors(): void {
    // Request interceptor
    this.axiosInstance.interceptors.request.use(
      async (config) => {
        const token = await this.authService.getIdToken();
        if (token) {
          config.headers.Authorization = `Bearer ${token}`;
        }
        console.log(`[API] ${config.method?.toUpperCase()} ${config.url}`);
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Response interceptor
    this.axiosInstance.interceptors.response.use(
      (response) => {
        console.log(`[API] Response: ${response.status}`);
        return response;
      },
      (error: AxiosError) => {
        console.log(`[API] Error: ${error.message}`);
        return Promise.reject(error);
      }
    );
  }

  async healthCheck(): Promise<any> {
    try {
      const response = await this.axiosInstance.get('/health');
      return response.data;
    } catch (error) {
      throw error;
    }
  }

  async getStudents(): Promise<any[]> {
    try {
      const response = await this.axiosInstance.get('/api/students');
      return response.data;
    } catch (error) {
      throw error;
    }
  }

  async getCourses(): Promise<any[]> {
    try {
      const response = await this.axiosInstance.get('/api/courses');
      return response.data;
    } catch (error) {
      throw error;
    }
  }

  async recordAttendance(
    studentId: string,
    courseId: string,
    isPresent: boolean,
    photoUrl?: string
  ): Promise<any> {
    try {
      const response = await this.axiosInstance.post('/api/attendance/record', {
        student_id: studentId,
        course_id: courseId,
        is_present: isPresent,
        photo_url: photoUrl,
        timestamp: new Date().toISOString(),
      });
      return response.data;
    } catch (error) {
      throw error;
    }
  }

  async getStatistics(courseId: string): Promise<any> {
    try {
      const response = await this.axiosInstance.get(
        `/api/statistics/${courseId}`
      );
      return response.data;
    } catch (error) {
      throw error;
    }
  }

  async uploadFile(file: File, uploadPath: string): Promise<string> {
    try {
      const formData = new FormData();
      formData.append('file', file);
      formData.append('path', uploadPath);

      const response = await this.axiosInstance.post('/api/upload', formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });

      return response.data.url;
    } catch (error) {
      throw error;
    }
  }
}

File: attendance_app/lib/services/api_client.dart
import 'dart:io';
import 'package:dio/dio.dart';

/// API Client for backend communication
class ApiClient {
  static final ApiClient _instance = ApiClient._internal();
  late Dio _dio;

  String baseUrl = 'http://localhost:8000'; // Change to your backend URL

  factory ApiClient() {
    return _instance;
  }

  ApiClient._internal() {
    _initializeDio();
  }

  void _initializeDio() {
    _dio = Dio(
      BaseOptions(
        baseUrl: baseUrl,
        connectTimeout: const Duration(seconds: 10),
        receiveTimeout: const Duration(seconds: 10),
        sendTimeout: const Duration(seconds: 10),
        contentType: 'application/json',
      ),
    );

    // Add interceptors
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          debugPrint('[API] ${options.method} ${options.path}');
          return handler.next(options);
        },
        onResponse: (response, handler) {
          debugPrint('[API] Response: ${response.statusCode}');
          return handler.next(response);
        },
        onError: (error, handler) {
          debugPrint('[API] Error: ${error.message}');
          return handler.next(error);
        },
      ),
    );
  }

  /// Add auth token to requests
  void setAuthToken(String token) {
    _dio.options.headers['Authorization'] = 'Bearer $token';
  }

  /// Health check
  Future<dynamic> healthCheck() async {
    try {
      final response = await _dio.get('/health');
      return response.data;
    } catch (e) {
      rethrow;
    }
  }

  /// Get all students
  Future<List<dynamic>> getStudents() async {
    try {
      final response = await _dio.get('/api/students');
      return response.data as List<dynamic>;
    } catch (e) {
      rethrow;
    }
  }

  /// Record attendance
  Future<dynamic> recordAttendance({
    required String studentId,
    required String courseId,
    required bool isPresent,
    String? photoUrl,
  }) async {
    try {
      final response = await _dio.post(
        '/api/attendance/record',
        data: {
          'student_id': studentId,
          'course_id': courseId,
          'is_present': isPresent,
          'photo_url': photoUrl,
          'timestamp': DateTime.now().toIso8601String(),
        },
      );
      return response.data;
    } catch (e) {
      rethrow;
    }
  }

  /// Upload file
  Future<String> uploadFile({
    required File file,
    required String uploadPath,
  }) async {
    try {
      final formData = FormData.fromMap({
        'file': await MultipartFile.fromFile(
          file.path,
          filename: file.path.split('/').last,
        ),
      });

      final response = await _dio.post(
        '/api/upload',
        data: formData,
        options: Options(
          contentType: 'multipart/form-data',
        ),
      );

      return response.data['url'] ?? '';
    } catch (e) {
      rethrow;
    }
  }

  /// Get student statistics
  Future<dynamic> getStatistics(String courseId) async {
    try {
      final response = await _dio.get('/api/statistics/$courseId');
      return response.data;
    } catch (e) {
      rethrow;
    }
  }
}
